DOCSP-41887 Async Migration Guide (#103)
mongodb/docs-pymongo: source/pymongo-async-migration.txt

.. _pymongo-async-migration:

===================================
Migrate to the {+driver-async+} Driver
===================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: motor, async, refactor, migration

Overview
--------

> The {+driver-async+} driver is a unification of {+driver-short+} and the `Motor
> library <https://www.mongodb.com/docs/drivers/motor/>`__. In this guide, you can

Reviewer: ```suggestion
The {+driver-async+} driver is a unification of {+driver-short+} and the `Motor
library <https://www.mongodb.com/docs/drivers/motor/>`__, which was the asynchronous 
MongoDB driver for Python. In this guide, you can
```

identify the changes you must make to migrate an application from
{+driver-short+} or Motor to the {+driver-async+} driver.

Migrate From {+driver-short+}
--------------------

The {+driver-async+} driver behaves similarly to {+driver-short+}, but
all methods that perform network operations are coroutines and must be awaited.
To migrate from {+driver-short+} to {+driver-async+}, you must update your code
to replace all uses of ``MongoClient`` with ``AsyncMongoClient`` and add the
``await`` keyword to all asynchronous method calls. If an asynchronous method is

> called within a function, you must also mark the function as ``async``.

Reviewer: S: convert to a list: eg

.. you must update your code in the following ways:

- Replace all uses of ...
- Add the ``await`` keyword to..
- ....

> 

Reviewer: S: add a section that introduces the subsections. eg The following sections describe how to implement the async API

Asynchronous Methods
~~~~~~~~~~~~~~~~~~~~

> The following methods are asynchronous in {+driver-async+}:

Reviewer: S: this sentence doesnt say much atm - maybe it can be a bit more descriptive about exactly what this section describes.

          await collection.drop_search_index(...)

   * - ``update_search_index()``
     - .. code-block:: python

          await collection.update_search_index(...)
   
   * - ``options()``
     - .. code-block:: python

          options = await collection.options()
   
   * - ``aggregate()``
     - .. code-block:: python

          async for doc in await collection.aggregate(...):
          ...

   * - ``aggregate_raw_batches()``
     - .. code-block:: python

          async for batch in await collection.aggregate_raw_batches(...):
          ...

   * - ``rename()``
     - .. code-block:: python

          await collection.rename(...)

   * - ``distinct()``
     - .. code-block:: python

          distinct_values = await collection.distinct(...)

   * - ``find_one_and_delete()``
     - .. code-block:: python

          result = await collection.find_one_and_delete(...)

   * - ``find_one_and_replace()``
     - .. code-block:: python

          result = await collection.find_one_and_replace(...)

   * - ``find_one_and_update()``
     - .. code-block:: python

          result = await collection.find_one_and_update(...)

Migrate From Motor
------------------

> The {+driver-async+} driver behaves nearly identically to the Motor library. In
> most cases, you can directly migrate existing Motor applications to
> {+driver-short+}, changing only the application's import statements, and
> by using ``AsyncMongoClient`` in place of ``MotorClient``.
> 
> The following section shows the method signature changes that you must make when
> migrating from Motor to the {+driver-async+} driver:

Reviewer: ```suggestion
The {+driver-async+} driver behaves nearly identically to the Motor library. In
most cases, you can directly migrate existing Motor applications to
{+driver-short+}, changing only the application's import statements.
The following code shows the difference in imports to use an XXX in Motor compared to Pymongo async:

.. code-block:: python
   
   # Motor Imports for zzz
   xxxx
   
   # PyMongo Async imports for zzz
   xxxx
   
You must also use the xxx type to create a client instead of MotorClient

The following section describes the method signature changes that you must make when
migrating from Motor to the {+driver-async+} driver.
```

.. _pymongo-async-migration:

===================================
Migrate to the {+driver-async+} Driver
===================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: motor, async, refactor, migration

Overview
--------

The {+driver-async+} driver is a unification of {+driver-short+} and the `Motor
library <https://www.mongodb.com/docs/drivers/motor/>`__. In this guide, you can

> identify the changes you must make to migrate an application from

Reviewer 2: Same here, the word "migrate" seems a little confusing for existing PyMongo users. Is there a different wording we can use to make sure users understand that they only need to make changes if they want to use the new async API?

Author: Reworded this after splitting pages

{+driver-short+} or Motor to the {+driver-async+} driver.

Migrate From {+driver-short+}
--------------------

The {+driver-async+} driver behaves similarly to {+driver-short+}, but
all methods that perform network operations are coroutines and must be awaited.
To migrate from {+driver-short+} to {+driver-async+}, you must update your code
in the following ways:

- Replace all uses of ``MongoClient`` with ``AsyncMongoClient``.
- Add the ``await`` keyword to all asynchronous method calls.
- If an asynchronous method is called within a function, mark the function as ``async``.

> The following sections describe how to implement the asynchronous API.

Reviewer 2: I'd also link to the official [Python async docs here](https://docs.python.org/3/library/asyncio.html) for some additional context.

Asynchronous Methods
~~~~~~~~~~~~~~~~~~~~

The following tables list the asynchronous methods that are available in the
{+driver-async+} driver. To call these methods, you must ``await`` them and call them
inside an ``async`` function.

Client Methods
``````````````

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Method
     - Example

   * - ``AsyncMongoClient()``
     - .. code-block:: python

          from pymongo import AsyncMongoClient

>           AsyncMongoClient(...)

Reviewer 2: To use an `AsyncMongoClient` as a context manager users will have to do `async with AsyncMongoClient(...)` instead of `with MongoClient(...)`.

   * - ``drop_search_index()``
     - .. code-block:: python

          await collection.drop_search_index(...)

   * - ``update_search_index()``
     - .. code-block:: python

          await collection.update_search_index(...)
   
   * - ``options()``
     - .. code-block:: python

          await collection.options()
   
   * - ``aggregate()``
     - .. code-block:: python

          async for doc in await collection.aggregate(...):
             ...

   * - ``aggregate_raw_batches()``
     - .. code-block:: python

          async for batch in await collection.aggregate_raw_batches(...):
             ...

   * - ``rename()``
     - .. code-block:: python

          await collection.rename(...)

   * - ``distinct()``
     - .. code-block:: python

          await collection.distinct(...)

   * - ``find_one_and_delete()``
     - .. code-block:: python

          await collection.find_one_and_delete(...)

   * - ``find_one_and_replace()``
     - .. code-block:: python

          await collection.find_one_and_replace(...)

   * - ``find_one_and_update()``
     - .. code-block:: python

          await collection.find_one_and_update(...)

> Migrate From Motor

Reviewer 3: I'd like to see the motor -> async pymongo migration as its own dedicated page. The pymongo sync -> async differences clutters the page for motor users.

Author: Split the pages into two

------------------

> The {+driver-async+} driver behaves nearly identically to the Motor library. In

Reviewer 3: This is a little misleading. Async PyMongo is (or should be) vastly superior to Motor in behavior both in terms of latency and throughput because Async Pymongo directly uses asyncio whereas Motor delegates all work to a thread pool.

Author: Updated the wording here, let me know if that works better!

.. _pymongo-async-migration:

===================================
Migrate to the {+driver-async+} Driver
===================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: motor, async, refactor, migration

Overview
--------

In September 2024, MongoDB released the {+driver-async+} driver to unify {+driver-short+}
and `Motor <https://www.mongodb.com/docs/drivers/motor/>`__, the asynchronous
MongoDB driver for Python. In this guide, you can identify the changes you must
make to migrate an application from {+driver-short+} or Motor to the
{+driver-async+} driver.

Migrate From {+driver-short+}
--------------------

The {+driver-async+} driver behaves similarly to {+driver-short+}, but
all methods that perform network operations are coroutines and must be awaited.
To migrate from {+driver-short+} to {+driver-async+}, you must update your code
in the following ways:

- Replace all uses of ``MongoClient`` with ``AsyncMongoClient``.

> - Add the ``await`` keyword to all asynchronous method calls .

Reviewer: ```suggestion
- Add the ``await`` keyword to all asynchronous method calls.
```

- If an asynchronous method is called within a function, mark the function as ``async``.

The following sections describe how to implement the asynchronous API.

Asynchronous Methods
~~~~~~~~~~~~~~~~~~~~

The following tables list the asynchronous methods that are available in the
{+driver-async+} driver. To call these methods, you must ``await`` them and call them
inside an ``async`` function.

Client Methods
``````````````

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Method
     - Example

   * - ``AsyncMongoClient()``
     - .. code-block:: python

          from pymongo import AsyncMongoClient

>           client = AsyncMongoClient(...)

Reviewer: S: perhaps this should have its own subsection as its a constructor rather than a method?

Author: I thought about that at first, but it seemed weird to be in its own table and it's related to the client so I figured it made sense here. Can change if anyone has strong feelings the other way though

   * - ``watch()``
     - .. code-block:: python

>           async with await client.watch(...) as stream:
>           ...

Reviewer: ```suggestion
          async with await client.watch(...) as stream:
              ...
```

   * - ``server_info()``
     - .. code-block:: python

>           info = await client.server_info(...)

Reviewer: Q: did you only assign values for methods that have a return type? I wonder if it would be more simple to just omit the assignment variable from all of the examples

Author: I took this from a list of methods that were probably all taken from random areas so I don't think there was an exact methodology. I removed them to make it simpler

   * - ``list_databases()``
     - .. code-block:: python

          databases = await client.list_databases()

   * - ``list_database_names()``
     - .. code-block:: python

          database_names = await client.list_database_names()

   * - ``drop_database()``
     - .. code-block:: python

          await client.drop_database(...)

Database Methods
````````````````

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Method
     - Example

   * - ``watch()``
     - .. code-block:: python

>           async with await db.watch(...) as stream:
>           ...

Reviewer: ```suggestion
          async with await db.watch(...) as stream:
              ...
```

   * - ``create_collection()``
     -   .. code-block:: python

          await db.create_collection(...)

   * - ``aggregate()``
     - .. code-block:: python

>           async with await client.admin.aggregate(...) as cursor:
>           ...

Reviewer: ```suggestion
          async with await client.admin.aggregate(...) as cursor:
              ...
```

          await collection.rename(...)

   * - ``distinct()``
     - .. code-block:: python

          distinct_values = await collection.distinct(...)

   * - ``find_one_and_delete()``
     - .. code-block:: python

          result = await collection.find_one_and_delete(...)

   * - ``find_one_and_replace()``
     - .. code-block:: python

          result = await collection.find_one_and_replace(...)

   * - ``find_one_and_update()``
     - .. code-block:: python

          result = await collection.find_one_and_update(...)

Migrate From Motor
------------------

The {+driver-async+} driver behaves nearly identically to the Motor library. In
most cases, you can directly migrate existing Motor applications to
{+driver-async+} by using ``AsyncMongoClient`` in place of ``MotorClient``, and
changing the application's import statements to import from ``pymongo``.

The following example shows the difference in imports to use a client for
read and write operations in Motor compared to {+driver-async+}:

.. code-block:: python

   # Motor client import
   from motor.motor_asyncio import AsyncIOMotorClient

   # {+driver-async+} client import
   from pymongo import AsyncMongoClient

> The following section shows the method signature changes that you must make when
> migrating from Motor to the {+driver-async+} driver.

Reviewer: ```suggestion
The following section shows the method signature changes that you must implement in your application
when migrating from Motor to the {+driver-async+} driver.
```
